### PR TITLE
8228609: G1 copy cost prediction uses used vs. actual copied bytes

### PR DESCRIPTION
**Pull request by linade**
I would like to backport 8228609 which fixes a prediction error regarding cost per bytes copied.

The patch does not apply cleanly, so this backport is basically changing the calculation of copied bytes.

https://bugs.openjdk.java.net/browse/JDK-8227442 is not in 11u. So this patch also uses G1ParScanThreadState::surviving_young_words() in place of _surviving_young_words to adjust for the correct cset index.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8228609](https://bugs.openjdk.java.net/browse/JDK-8228609): G1 copy cost prediction uses used vs. actual copied bytes

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk11u-dev pull/927/head:pull/927` \
`$ git checkout pull/927`

Update a local copy of the PR: \
`$ git checkout pull/927` \
`$ git pull https://git.openjdk.java.net/jdk11u-dev pull/927/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 927`

View PR using the GUI difftool: \
`$ git pr show -t 927`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk11u-dev/pull/927.diff">https://git.openjdk.java.net/jdk11u-dev/pull/927.diff</a>

</details>

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
This PR is not recognized as backport. Please fix that (e.g. by changing the title to "Backport <hash of original commit>").

**Comment by linade** [comment]
@GoeLin I enabled the tests. Some of them seem to be blocked by a gcc dependency issue. Would you take a look?

**Comment by GoeLin** [comment]
Hi, you should only label a change jdk11u-fix-request once it is ready for integration. I'll remove the label, please add it again if you have a review.

**Comment by linade** [comment]
Gentle ping. Can anyone review this?

The reason I believe it's worth a backport is, apart from what's mentioned in https://bugs.openjdk.java.net/browse/JDK-8228609, it correctly accounts for early reclaimed humongous objects. I think originally the oversight of the size of early reclaimed humongous objects cause the calculation to be erroneous, resulting in a very high copy time prediction and unstable gc frequency.

File: src/hotspot/share/gc/g1/g1Policy.cpp
Comment (TheRealMDoerr) [inline src/hotspot/share/gc/g1/g1Policy.cpp:669]: `freed_bytes` is now unused.

**Review by TheRealMDoerr — COMMENTED** [review]
This change is very different from the original one, but fixing it in 11u sounds feasible to me. We'll test it.
Do you have a specific test case which allows observing unstable gc frequency?

**Comment by linade** [comment]
> This change is very different from the original one, but fixing it in 11u sounds feasible to me. We'll test it. Do you have a specific test case which allows observing unstable gc frequency?

We observed this in an application. With additional logging, we can observe the miscalculation from, e.g., test/hotspot/jtreg/gc/g1/TestEagerReclaimHumongousRegions.java

The current way to get copied_bytes will result in negative value. I had
```
..
[0.708s][error][gc] GC(186) cset.used=14641416, freed_bytes=40855816, copied_bytes=-26214400, actual_copied_bytes=0
[0.714s][error][gc] GC(188) cset.used=14641416, freed_bytes=40855816, copied_bytes=-26214400, actual_copied_bytes=0
[0.720s][error][gc] GC(190) cset.used=14641416, freed_bytes=40855816, copied_bytes=-26214400, actual_copied_bytes=0
[0.726s][error][gc] GC(192) cset.used=14641416, freed_bytes=40855816, copied_bytes=-26214400, actual_copied_bytes=0
[0.732s][error][gc] GC(194) cset.used=14641416, freed_bytes=40855816, copied_bytes=-26214400, actual_copied_bytes=0
[0.738s][error][gc] GC(196) cset.used=14641416, freed_bytes=40855816, copied_bytes=-26214400, actual_copied_bytes=0
[0.744s][error][gc] GC(198) cset.used=14641416, freed_bytes=40855816, copied_bytes=-26214400, actual_copied_bytes=0
```
copied_bytes is negative because it is deduced from ``size_t copied_bytes = _collection_set->bytes_used_before() - freed_bytes;``, and ``_collection_set->bytes_used_before()`` didn't consider eager-reclaimed humongous, so it appears to be much smaller than ``freed_bytes``.

There currently is an if condition ``if (_collection_set->bytes_used_before() > freed_bytes)`` to prevent we from getting a negative value. So the current code simply doesn't report the data to the analytics.

Not reporting is fine. At worst we lose some data points. But it can also happen that we get a positive, near-zero ``copied_bytes``. It will make the value ``double cost_per_byte_ms = average_copy_time / (double) copied_bytes;`` very big. This causes young generation to shrink.

File: src/hotspot/share/gc/g1/g1Policy.cpp
Comment (linade) [reply src/hotspot/share/gc/g1/g1Policy.cpp]: Removed. Thanks.

**Review by TheRealMDoerr — APPROVED** [review]
Looks technically good and the fix makes sense to me. Our nightly tests haven't found any issues related to it. It extracts the essential fix from the original change which differs a lot: https://github.com/openjdk/jdk/commit/9a4c25731eef42d18ba0c20ff3eb42f32ae90c00
An additional person with GC experience should review it.
Approvers will have to judge how relevant it is. A comment from the original author would be helpful.

**Comment by rkennke** [comment]
I *hink* it makes sense. It is not exactly a backport, but rather a re-write, and I find it difficult to follow and convince myself that it's doing the correct thing. Therefore it seems somewhat risky.

File: src/hotspot/share/gc/g1/g1ParScanThreadState.cpp
Comment (rkennke) [inline src/hotspot/share/gc/g1/g1ParScanThreadState.cpp:94]: Was that code using the wrong index before? IOW, it used the surviving_young_words of the wrong regions?

Comment (linade) [reply src/hotspot/share/gc/g1/g1ParScanThreadState.cpp:94]: Yes, that's what I think.

So this pr is achieving two things (sorry if it were not clear):

1. fix the index issue: this causes calculation of the total surviving young words to be incorrect. The correct stats are stored in _surviving_young_words[1, .., n] instead of [0, .., n-1]
3. use a precise "copied bytes" instead of inferring from (used() - freed())

**Pull request closed without merging**
